### PR TITLE
Allow applications to request exlcusive access to hosts

### PR DESCRIPTION
**Pull request by bratseth**
This allows applications to specify they want to be the only tenant running on their docker hosts to provide another layer of security for applications with sensitive data.

I finally removed support for cluster membership serialized forms without a group setting. That has not been in use for more than 1.5 years. Many of the changes are about that.

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java [hunk 1]
```diff
@@ -127,6 +143,33 @@ private boolean offeredNodeHasParentHostnameAlreadyAccepted(Collection<Prioritiz
         return false;
     }
 
+    /**
+     * If a parent host is given, and it hosts another application which requires exclusive access to the physical
+     * host, then we cannot host this application on it.
+     */
+    private boolean hostCanHost(ApplicationId application, Optional<String> parentHostname) {
```
Comment (mpolden) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java]: Why is this needed? Isn't the same requirement satisfied by `hostHasOnly`? In any case the methods could be collapsed to share some code.

**Review by mpolden — COMMENTED** [review]
One comment, otherwise LGTM.

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java (hunk 1 above)
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java]: It is not the same requirement - one is that there is another *exclusive* application on the host, the other is that there is another application on the host. The need for both stems from the two cases we need to handle: We are the exclusive app and is offered a node, and we are not exclusive but are offered a node on a host where some other tenant is exclusive.

I could combine them and pass a parameter about whether to also check for exclusiveness, but imho it is not worth it because it is hard to read that kind of convoluted tests.

Btw, perhaps we should change the rule to be *same tenant* vs *same application*?

Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java]: I decided to switch to exclusivity on the tenant level.

Comment (mpolden) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java]: I see, thanks for explaining. Maybe the naming could be improved? E.g. `alreadyExclusivelyAllocated` and `canAllocateExclusively`?

Agree that tenant granularity would be better, would provide basically the same security isolation while nodes can be used more effectively.

**Comment by bratseth** [comment]
Changed to "exclusiveTo(tenant, hostname)" and "hostsOnly(tenant, hostname)". 
Better?

**Comment by mpolden** [comment]
Yes. 🙂
